### PR TITLE
[docs] document how to use a custom objective function via the C API

### PR DESCRIPTION
**Pull request by parsiad**
Follow-up to #6393.

Clarifies that `objective=custom` must be provided via `parameters` in the C API if you want to use a custom objective function and provide gradients/hessians manually.

### CODE REVIEW

**Comment by parsiad** [comment]
@microsoft-github-policy-service agree

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -158,6 +158,9 @@ struct Config {
   // descl2 = ``rank_xendcg``, `XE_NDCG_MART <https://arxiv.org/abs/1911.09798>`__ ranking objective function, aliases: ``xendcg``, ``xe_ndcg``, ``xe_ndcg_mart``, ``xendcg_mart``
   // descl2 = ``rank_xendcg`` is faster than and achieves the similar performance as ``lambdarank``
   // descl2 = label should be ``int`` type, and larger number represents the higher relevance (e.g. 0:bad, 1:fair, 2:good, 3:perfect)
+  // desc = custom
```
Comment (jameslamb) [inline include/LightGBM/config.h:161]:
```suggestion
  // desc = custom objective function (gradients and hessians not computed directly by LightGBM)
```

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -158,6 +158,9 @@ struct Config {
...
+  // descl2 = ``custom``
+  // descl2 = **Note**: Not supported in CLI version; must be set explicitly when using the C API (for Python and R APIs, pass the custom objective directly into the train method)
```
Comment (jameslamb) [inline include/LightGBM/config.h:163]:
```suggestion
  // descl2 = **Note**: Not supported in CLI version; must be passed through parameters explicitly
```

You're right that this won't be supported in the CLI. But let's please make this a bit more generic... there are other libraries that use LightGBM's C API besides the R and Python packages (like the Java and .NET interfaces), and there are other interfaces not named "train" (like `update()`, `fit()`, and `cv()` in the Python API).

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this!

I've updated the title and added a PR description. In the future when you contribute here, please do provide a PR description briefly describing what the PR does and how it improves the project, including links to any other relevant information. This helps maintainers to understand how what the goal is (to provide useful feedback) and helps other people finding the PR in the future (e.g. from search engines or release notes).

After applying the suggestions I left, please run the script we use to automatically generate parameter documentation and the corresponding C++ code.

```shell
python ./helpers/parameter_generator.py
```

**Review by jameslamb — APPROVED** [review]
Thanks very much! I just pushed commits to get this up to date with the latest changes in this repo (now that we've fixed continuous integration), and to split up 2 lines of documentation.

I also changed this from your phrasing:

> *must be passed through parameters explicitly*

to this

> *must be passed through parameters explicitly in the C API*

So that users of the R and Python packages don't get confused and think they need to pass `"objective": "custom"` (they do not).

👋🏻 Come back and contribute any time!
